### PR TITLE
[SYCL] Improve memobj creation for devices w/o host unified memory

### PR DESCRIPTION
**Pull request by sergey-semenov**
The aggressive use of USE_HOST_PTR memory object creation option led to
significant performance overhead on devices without host unified memory
support. Furthermore, any discard access modes were ignored during the
allocation leading to unnecessary copy.

This patch makes it so that allocations for such devices are made with
COPY_HOST_PTR instead, and only if necessitated by the access mode.
These allocations are also not linked anymore, since linked allocations
rely on USE_HOST_PTR option.

### CODE REVIEW

**Comment by sergey-semenov** [comment]
TBD: add tests

File: sycl/source/detail/memory_manager.cpp [hunk 1]
```diff
@@ -137,16 +137,39 @@ void *MemoryManager::allocateInteropMemObject(
...
+      Result |= PI_MEM_FLAGS_HOST_PTR_COPY;
+    else {
+      // Create the memory object using the host pointer only if the device
+      // supports host_unified_memory to avoid potential copy overhead.
+      // TODO This check duplicates the one performed in the GraphBuilder during
+      // AllocaCommand creation. This information should be propagated here
+      // instead, which would be a breaking ABI change.
+      bool HostUnifiedMemory = true;
+      for (device Device : TargetContext->getDevices()) {
+        if (!Device.get_info<info::device::host_unified_memory>())
```
Comment (romanovvlad) [inline sycl/source/detail/memory_manager.cpp:155]:
```suggestion
      for (const device &Device : TargetContext->getDevices()) {
        HostUnifiedMemory &= Device.get_info<info::device::host_unified_memory>();
```

File: sycl/source/detail/memory_manager.cpp [hunk 2]
```diff
@@ -137,16 +137,39 @@ void *MemoryManager::allocateInteropMemObject(
   return UserPtr;
 }
 
+RT::PiMemFlags getMemObjCreationFlags(ContextImplPtr TargetContext,
```
Comment (romanovvlad) [inline sycl/source/detail/memory_manager.cpp:140]:
```suggestion
RT::PiMemFlags getMemObjCreationFlags(const ContextImplPtr &TargetContext,
```

File: sycl/source/detail/scheduler/graph_builder.cpp [hunk 3]
```diff
@@ -574,6 +605,14 @@ Scheduler::GraphBuilder::findAllocaForReq(MemObjRecord *Record,
   return (Record->MAllocaCommands.end() != It) ? *It : nullptr;
 }
 
+static bool checkHostUnifiedMemory(ContextImplPtr Ctx) {
```
Comment (romanovvlad) [inline sycl/source/detail/scheduler/graph_builder.cpp:608]:
```suggestion
static bool checkHostUnifiedMemory(const ContextImplPtr &Ctx) {
```

**Review by romanovvlad — CHANGES_REQUESTED** [review]

**Comment by sergey-semenov** [comment]
Note: there's a non-breaking ABI change in the latest version of the PR, a new symbol is added to `SYCLMemObjT`.

File: sycl/source/detail/scheduler/graph_builder.cpp [hunk 4]
```diff
@@ -69,6 +69,37 @@ static bool isAccessModeAllowed(access::mode Required, access::mode Current) {
   return false;
 }
 
+/// Combines two access modes into a single one that allows both.
+static access::mode combineAccessModes(access::mode A, access::mode B) {
+  if (A == B)
```
Comment (romanovvlad) [inline sycl/source/detail/scheduler/graph_builder.cpp:74]: What about:

{
  if(A == B)
    return A;

  if(A == access::mode::discard_write && (B == access::mode::discard_read_write || B == access::mode::write))
      return B;

  if(B == access::mode::discard_write && (A == access::mode::discard_read_write || A == access::mode::write))
      return A;

  return access::mode::read_write;
}

?

File: sycl/source/detail/scheduler/graph_builder.cpp [hunk 5]
```diff
@@ -574,6 +605,14 @@ Scheduler::GraphBuilder::findAllocaForReq(MemObjRecord *Record,
   return (Record->MAllocaCommands.end() != It) ? *It : nullptr;
 }
 
+static bool checkHostUnifiedMemory(const ContextImplPtr &Ctx) {
+  for (device Device : Ctx->getDevices()) {
```
Comment (romanovvlad) [inline sycl/source/detail/scheduler/graph_builder.cpp:609]:
```suggestion
  for (const device &Device : Ctx->getDevices()) {
```

File: sycl/source/detail/scheduler/graph_builder.cpp [hunk 6]
```diff
@@ -618,13 +667,22 @@ AllocaCommandBase *Scheduler::GraphBuilder::getOrCreateAllocaForReq(
...
-              findAllocaForReq(Record, Req, Record->MCurContext);
-
-          // Cannot setup link if candidate is linked already
-          if (LinkedAllocaCmdCand && !LinkedAllocaCmdCand->MLinkedAllocaCmd)
-            LinkedAllocaCmd = LinkedAllocaCmdCand;
+          // Linked commands assume that the host allocation is reused by the
+          // plugin runtime and that can lead to unnecessary copy overhead on
+          // devices that do not support host unified memory. Do not link the
+          // allocations in this case.
+          ContextImplPtr NonHostCtx = Queue->is_host()
```
Comment (romanovvlad) [inline sycl/source/detail/scheduler/graph_builder.cpp:674]:
```suggestion
          const ContextImplPtr &NonHostCtx = Queue->is_host()
```

**Review by romanovvlad — COMMENTED** [review]
LGTM

File: sycl/source/detail/scheduler/graph_builder.cpp (hunk 4 above)
Comment (sergey-semenov) [reply sycl/source/detail/scheduler/graph_builder.cpp:74]: That looks much cleaner indeed, applied.

**Review by romanovvlad — APPROVED** [review]

**Pull request merged**